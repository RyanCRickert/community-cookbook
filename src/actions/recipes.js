import uuid from "uuid";
import database from "../firebase/firebase";

export const addRecipe = (recipe) => ({
  type: "ADD_RECIPE",
  recipe
});

export const startAddRecipe = (recipeData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      name = "",
      ingredients = [],
      instructions = "",
      cookTime = 0,
      createdAt = 0
  } = recipeData;
    const recipe = { name, instructions, createdAt, ingredients, cookTime };

    return database.ref(`recipes`).push(recipe).then((ref) => {
      dispatch(addRecipe({
        id: ref.key,
        ...recipe
      }));
    });
  };
};

export const addIngredient = (id, ingredient) => ({
  type: "ADD_INGREDIENT",
  id,
  ingredient
});

export const startAddIngredient = (id, ingredient) => {
  return (dispatch) => {
    const {
      name = ""
    } = ingredientData;
    const ingredient = {name};

    return database.ref(`recipes/${id}/ingredients`).push(ingredient).then((ref) => {
      dispatch(addIngredient({
        iid: ref.key,
        ingredient
      }));
    });
  };
};

export const editIngredient = (id, iid, update) => {
  type:"EDIT_INGREDIENT",
  id,
  iid,
  update
};

export const startEditIngredient = (id, iid, update) => {
  return (dispatch) => {
    return database.ref(`recipes/${id}/ingredients${iid}`)
  };
};

export const removeRecipe = ({ id } = {}) => ({
  type: "REMOVE_RECIPE",
  id
});

export const startRemoveRecipe = ({ id }) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    return database.ref(`recipes/${id}`).remove().then(() => {
      dispatch(removeRecipe({ id }));
    });
  };
};

export const editRecipe = (id, updates) => ({
  type: "EDIT_RECIPE",
  id,
  updates
});

export const startEditRecipe = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    return database.ref(`recipes/${id}`).update(updates).then(() => {
      dispatch(editRecipe(id, updates));
    });
  };
};

export const setRecipes = (recipes) => ({
  type: "SET_RECIPES",
  recipes
});

export const startSetRecipes = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid

    return database.ref(`recipes`).once("value", (snapshot) => {
      const recipes = [];

      snapshot.forEach((childSnapshot) => {
        recipes.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setRecipes(recipes));
    });
  };
};

export const setIngredients = (id, ingredients) => ({
  type: "SET_INGREDIENTS",
  id,
  ingredients
});

export const startSetIngredients = () => {
  return (dispatch) => {
    return database.ref(`recipes/${id}/ingredients`).once("value", (snapshot) => {
      const ingredients = [];

      snapshot.forEach((childSnapshot) => {
        recipes.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setIngredients(ingredients));
    });
  };
};