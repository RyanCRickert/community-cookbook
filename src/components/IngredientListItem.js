import React from "react";

export default (props) => (
  <div className="ingredient-item">
    <div className="ingredient-item__left">
      {props.name}
    </div>
    <div
      className="ingredient-item__right"
      onClick={(e) => {
        props.handleRemoveIngredient(props.name)
      }}>
      Remove   
    </div>
  </div>
)