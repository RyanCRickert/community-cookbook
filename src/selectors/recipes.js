import moment from "moment";

export default (recipes, {name, sortBy, startDate, endDate}) => {
    return recipes.filter((recipe) => {
        const createdAtMoment = moment(recipe.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, "day") : true ;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, "day") : true ;
        const nameMatch = recipe.name.toLowerCase().includes(name.toLowerCase());

        return startDateMatch && endDateMatch && nameMatch;
    }).sort((a , b) => {
        if (sortBy === "date") {
            return a.createdAt < b.createdAt ? 1 : -1;
        } else if (sortBy === "name") {
            return a.name > b.name ? 1 : -1;
        } else if (sortBy === "cookTime") {
            return a.cookTime < b.cookTime ? 1 : -1;
        } else if (sortBy === "feeds") {
            return a.feeds < b.feeds ? 1 : -1;
        }
    });
};