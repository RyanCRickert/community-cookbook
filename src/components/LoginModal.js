import React from "react";
import Modal from "react-modal";
import { connect } from "react-redux";
import { firebase, googleAuthProvider } from "../firebase/firebase";
import { startGoogleLogin } from "../actions/auth";

export class LoginModal extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			email: "",
			password: "",
			newUser: false,
			error: ""
      
		}
	}

	onPasswordChange = (e) => {
		const password = e.target.value;
		this.setState(() => ({ password }));
	};

	onEmailChange = (e) => {
		const email = e.target.value;
    this.setState(() => ({ email }));
	};

	handleLogin = (e) => {
		e.preventDefault();

		firebase.auth().signInWithEmailAndPassword(this.state.email, this.state.password).catch((e) => {
			if(e.code === "auth/user-not-found") {
				this.setState({ error: "No user with that email found." });
			} else if (e.code === "auth/wrong-password") {
				this.setState({ error: "Incorrect password, please try again." });
			} else {
				this.setState({ error: e.message });
			}
    })
	};

	handleStartNewUser = () => {
		this.setState({ newUser: true });
	};

	handleCloseModal = () => {
		this.props.handleCloseModal()
		this.setState({ newUser: false });
	}

	handleCancelNewUser = () => {
		this.setState({ newUser: false });
	}

	handleCreateNewUser = (e) => {
		e.preventDefault();

		firebase.auth().createUserWithEmailAndPassword(this.state.email, this.state.password).catch((e) => {
			if (e.code === "auth/email-already-in-use") {
				this.setState({ error: "That email is already in use." });
			} else if (e.code === "auth/invalid-email") {
				this.setState({ error: "Incorrect username or password." });
			} else {
				this.setState({ error: e.message });
			}
		});
	}

	handleGoogleLogin = () => {
		startLogin();
	}

	render() {
		return (
			<Modal
				ariaHideApp={false}
				onRequestClose={this.handleCloseModal}
				isOpen={!!this.props.modalOpen}
				contentLabel="Login"
				closeTimeoutMS={200}
				className="modal"
			>
				{this.state.newUser ?
					<form className="form" onSubmit={this.handleCreateNewUser}>
					{this.state.error && <p className="modal__error">{this.state.error}</p>}
					<input
						className="text-input"
						type="text"
						placeholder="Email"
						autoFocus
						value={this.state.email}
						onChange={this.onEmailChange}
					/>
					<input
						className="text-input"
						placeholder="Password"
						type="password"
						value={this.state.password}
						onChange={this.onPasswordChange}
					></input>
					<div className="modal-button">
						<button className="button">Create</button>
					</div>
					<div>
						<button className="button" onClick={this.handleCancelNewUser}>Cancel</button>	
					</div>
					<div className="modal-button__margin-top">
						<button className="button" onClick={this.handleCloseModal}>Close</button>
					</div>
					</form>
					:
					<div>
						<form className="form" onSubmit={this.handleLogin}>
							{this.state.error && <p className="modal__error">{this.state.error}</p>}
							<input
								className="text-input"
								type="text"
								placeholder="Email"
								autoFocus
								value={this.state.email}
								onChange={this.onEmailChange}
							/>
							<input
								className="text-input"
								placeholder="Password"
								type="password"
								value={this.state.password}
								onChange={this.onPasswordChange}
							></input>
							<div className="modal-button">
								<div className="modal-button__group">
									<button className="button">Sign In</button>
									<button className="button" onClick={this.handleStartNewUser}>New User</button>
								</div>
								<div className="modal-google" onClick={(() => this.props.startGoogleLogin())}>Google Login</div>
							</div>
						</form>
						<div>
							<button className="button" onClick={this.handleCloseModal}>Close</button>
						</div>
					</div>
				}
      </Modal>
		)
	}
}

const mapDispatchToProps = (dispatch) => ({
	startGoogleLogin: () => dispatch(startGoogleLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginModal);