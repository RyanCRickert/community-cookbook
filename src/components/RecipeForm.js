import React from "react";
import moment from "moment";
import { Link } from "react-router-dom";
import IngredientListItem from "./IngredientListItem";

export default class RecipeForm extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			name: props.recipe ? props.recipe.name : "",
			instructions: props.recipe ? props.recipe.instructions : "",
			ingredients: props.recipe ? props.recipe.ingredients : [],
			cookTime: props.recipe ? parseInt(props.recipe.cookTime) : 1,
			feeds: props.recipe ? parseInt(props.recipe.feeds) : 1,
			category: props.recipe ? props.recipe.category : "",
			createdAt: props.recipe ? moment(props.recipe.createdAt) : moment(),
			error: ""
		}
	}

	onNameChange = (e) => {
		const name = e.target.value;
		this.setState(() => ({ name }));
	};

	onCookChange = (e) => {
		const cookTime = e.target.value;
		this.setState(() => ({ cookTime: parseInt(cookTime) }));
	};

	onFeedChange = (e) => {
		const feeds = e.target.value;
		this.setState(() => ({ feeds: parseInt(feeds) }));
	};

	onInstructionsChange = (e) => {
		const instructions = e.target.value;
		this.setState(() => ({ instructions }));
	};

	onCategoryChange = (e) => {
    if (e.target.value === "Beef") {
      this.setState({ category: "Beef"})
    } else if (e.target.value === "Chicken") {
      this.setState({ category: "Chicken"})
    } else if (e.target.value === "Other") {
      this.setState({ category: "Other"})
    } else if (e.target.value === "Meatless") {
      this.setState({ category: "Meatless"})
    } else if (e.target.value === "Pork") {
      this.setState({ category: "Pork"})
    }
  };

	handleRemoveIngredient = (ingredientToRemove) => {
		this.setState((prevState) => ({
			ingredients: prevState.ingredients.filter((ingredient) => ingredientToRemove !== ingredient)
		}));
	};

	handleAddIngredient = (e) => {
		e.preventDefault();

		const ingredient = e.target.elements.ingredient.value.trim();
		let error = false;

		if (!e.target.elements.ingredient.value) {
			this.setState({ error: "*Please enter a valid ingredient*" });
			error = true;
		} else if (this.state.ingredients.indexOf(ingredient) > -1) {
			this.setState({ error: "*This ingredient already exists*" });
			error = true;
		}

		if (!error) {
			e.target.elements.ingredient.value = "";
			this.setState((prevState) => ({ ingredients: prevState.ingredients.concat(ingredient) }));
			this.setState({ error: "" });
		}


	};

	onSubmit = (e) => {
		e.preventDefault();

		if (!this.state.name) {
			this.setState(() => ({ error: "Please enter a name." }));
		} else if (!this.state.ingredients) {
			this.setState(() => ({ error: "Please enter at least one ingredient." }));
		}  else if (!this.state.cookTime) {
			this.setState(() => ({ error: "Please verify the cook time." }));
		}  else if (!this.state.feeds) {
			this.setState(() => ({ error: "Please verify the feeds number." }));
		} else {
			this.setState(() => ({ error: "" }));
			this.props.onSubmit({
				name: this.state.name,
				ingredients: this.state.ingredients,
				createdAt: this.state.createdAt.valueOf(),
				instructions: this.state.instructions,
				cookTime: parseInt(this.state.cookTime),
				feeds: parseInt(this.state.feeds),
				category: this.state.category
			});
		}
	};

	render() {
		const pStyle = {marginTop: "0"};
		return (
			<div className="form-edit">
				{this.state.error && <p className="form__error">{this.state.error}</p>}
				<input
					className="text-input"
					type="text"
					placeholder="Name of Dish"
					autoFocus
					value={this.state.name}
					onChange={this.onNameChange}
				/>
				<div className="number-input">
				<div>
					Cook Time : 
					<input
						className="number-input__item"
						type="number"
						value={1}
						min={1}
						value={this.state.cookTime}
						onChange={this.onCookChange}
					/>
					&nbsp;{this.state.cookTime === 1 ? "minute" : "minutes"}
				</div>
				<div>
					Feeds : 
					<input
						className="number-input__item"
						type="number"
						value={1}
						min={1}
						value={this.state.feeds}
						onChange={this.onFeedChange}
					/>
					&nbsp;{this.state.feeds === 1 ? "person" : "people"}
				</div>
				<div>
					Category : 
				<select
					className="number-input__item-category"
					value={this.state.category}
					onChange={this.onCategoryChange}
				>
					<option value={null}></option>
					<option value="Beef">Beef</option>
					<option value="Chicken">Chicken</option>
					<option value="Meatless">Meatless</option>
					<option value="Other">Other</option>
					<option value="Pork">Pork</option>
				</select>
				</div>
				</div>
				<form onSubmit={this.handleAddIngredient}>
					<input
						className="text-input__ingredient"
						type="text"
						name="ingredient"
						placeholder="Add Ingredient"
						value={this.state.ingredient}
					/>
					<button className="button text-input__button">Add Ingredient</button>
				</form>
				{this.state.ingredients.length === 0 ?
					<p style={pStyle}>*Please add at least one ingredient*</p> :
					this.state.ingredients.map((ingredient) => (
						<IngredientListItem
							key={ingredient}
							name={ingredient}
							handleRemoveIngredient={this.handleRemoveIngredient}
						/>
					)
					)}
				<form onSubmit={this.onSubmit}>
					<textarea
						className="text-area"
						placeholder="Cooking instructions"
						value={this.state.instructions}
						onChange={this.onInstructionsChange}
					></textarea>
					<div>
						<button className="button">{this.props.actionType} Recipe</button>
					</div>
				</form>
				<Link to="/" className="button home-button">Home</Link>
			</div>
		)
	}
}