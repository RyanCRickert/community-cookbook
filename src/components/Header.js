import React from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import FontAwesome from "react-fontawesome";
import { startLogout } from "../actions/auth";
import { firebase } from "../firebase/firebase";
import LoginModal from "./LoginModal";

export class Header extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      loggedIn: null,
      modalOpen: false
    }

    this.handleOpenModal = this.handleOpenModal.bind(this);
    this.handleCloseModal = this.handleCloseModal.bind(this);
  }

  componentDidMount() {
    firebase.auth().onAuthStateChanged((user) => {
      if (user) {
        this.setState(() => ({ loggedIn: true }));
      } else {
        this.setState(() => ({ loggedIn: false }));
      }
    })
  }

  handleOpenModal() {
    this.setState(() => ({ modalOpen: true }));
  }

  handleCloseModal() {
    this.setState(() => ({ modalOpen: false }));
  }

  render() {
    return (
      <header className="header">
        <div className="content-container header-list">
          <div className="header-main">
            <h1 className="header-title" onClick={this.props.openWelcomeModal}>Community Cookbook</h1>
            <h3 className="header-subtitle">A place for sharing</h3>
          </div>
            {this.state.loggedIn
              ?
              <div className="header-buttons">
                <div className="header-log">
                  <Link className="button header-button" to="/add">New Recipe</Link>
                </div>
                <div className="header-log__fa">
                  <button className="button header-button" onClick={this.props.startLogout}>Logout</button>
                  <Link to="/cart"><div className="header-fa"><FontAwesome name="shopping-cart"/></div></Link>
                </div>
              </div>
              :
              <div className="header-log">
                <LoginModal modalOpen={this.state.modalOpen} handleCloseModal={this.handleCloseModal}/>
                <button className="button header-button header-button__login" onClick={this.handleOpenModal}>Login</button>
              </div>
            }
        </div>
      </header>
    )
  }

};

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header)