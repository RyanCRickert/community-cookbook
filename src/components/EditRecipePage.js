import React from "react";
import { connect } from "react-redux";
import RecipeForm from "./RecipeForm";
import { startEditRecipe, startRemoveRecipe } from "../actions/recipes";
import RemoveRecipeModal from "./RemoveRecipeModal";

export class EditRecipePage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      modalOpen: false
    }

    this.handleOpenModal = this.handleOpenModal.bind(this);
    this.handleCloseModal = this.handleCloseModal.bind(this);
  }
  onSubmit = (recipe) => {
    this.props.startEditRecipe(this.props.recipe.id, recipe);
    this.props.history.push("/");
  };

  onRemove = () => {
    this.props.startRemoveRecipe({id : this.props.recipe.id});
    this.props.history.push("/");
  }

  handleOpenModal() {
    this.setState(() => ({ modalOpen: true }));
  }

  handleCloseModal() {
    this.setState(() => ({ modalOpen: false }));
  }
    
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Recipe</h1>
          </div>
        </div>
          <div className="content-container">
            <RecipeForm
              actionType={"Edit"}
              recipe={this.props.recipe}
              onSubmit={this.onSubmit}
            />
          <button className="button button--secondary" onClick={this.handleOpenModal}>Remove Recipe</button>
          <RemoveRecipeModal
          modalOpen={this.state.modalOpen}
          handleCloseModal={this.handleCloseModal}
          onRemove={this.onRemove}
          recipeName={this.props.recipe.name}
          />
          </div>
        </div>
    )}
}

const mapStateToProps = (state, props) => ({
  recipe: state.recipes.find((recipe) => recipe.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  startEditRecipe: (id, recipe) => dispatch(startEditRecipe(id, recipe)),
  startRemoveRecipe: (data) => dispatch(startRemoveRecipe(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditRecipePage);