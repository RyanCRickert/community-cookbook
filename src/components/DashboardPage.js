import React from "react";
import Header from "./Header";
import RecipeList from "./RecipeList";
import RecipeListFilters from "./RecipeListFilters";
import { Link } from "react-router-dom";

export default () => (
  <div className="content-container">
    <Header />
    <RecipeListFilters />
    <RecipeList />
  </div>
);