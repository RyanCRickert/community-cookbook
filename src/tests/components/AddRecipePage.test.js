import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import { AddRecipePage } from "../../components/AddRecipePage";
import recipes from "../fixtures/recipes";

let startAddRecipe, history, wrapper;

beforeEach(() => {
  startAddRecipe = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<AddRecipePage
    buttonType={"Add"}
    startAddRecipe={startAddRecipe}
    history={history}
    />);
});

test("should render AddRecipePage correctly", () => {
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test("should handle onSubmit", () => {
  expect(toJSON(wrapper)).toMatchSnapshot();
  wrapper.find("RecipeForm").prop("onSubmit")(recipes[1]);
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(startAddRecipe).toHaveBeenLastCalledWith(recipes[1]);
  expect(toJSON(wrapper)).toMatchSnapshot();
});