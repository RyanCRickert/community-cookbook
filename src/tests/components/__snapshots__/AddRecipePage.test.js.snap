// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle onSubmit 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Add Recipe
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <RecipeForm
      actionType="Add"
      onSubmit={[Function]}
    />
  </div>
</div>
`;

exports[`should handle onSubmit 2`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Add Recipe
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <RecipeForm
      actionType="Add"
      onSubmit={[Function]}
    />
  </div>
</div>
`;

exports[`should render AddRecipePage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Add Recipe
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <RecipeForm
      actionType="Add"
      onSubmit={[Function]}
    />
  </div>
</div>
`;
